alpenhorn/io/nearline.py

```diff
@@ -0,0 +1,408 @@
+
+        # If it's restorable, restore it
+        if state == self._lfs.HSM_RELEASED:
+            self._lfs.hsm_restore(fullpath)
+
+        # Returns True if file is readable.
+        return state == self._lfs.HSM_RESTORED or state == self._lfs.HSM_UNARCHIVED
+
+    def ready_pull(self, req: ArchiveFileCopyRequest) -> None:
+        """Ready a file to be pulled as specified by `req`.
+
+        Parameters
+        ----------
+        req : ArchiveFileCopyRequest
+            the copy request to ready.  We are the source node (i.e.
+            `req.node_from == self.node`).
+        """
+        ready = self.ready_path(req.file.path)
+
+        # Update DB based on HSM state
+        ArchiveFileCopy.update(ready=ready).where(
+            ArchiveFileCopy.file == req.file, ArchiveFileCopy.node == self.node
+        ).execute()
+
+    def release_bytes(self, size: int) -> None:
+        """Does nothing."""
+        pass
+
+    def reserve_bytes(self, size: int, check_only: bool = False) -> bool:
+        """Returns True."""
+        return True
+
+
+class NearlineGroupIO(BaseGroupIO):
+    """Nearline Group I/O
+
+    The Nearline Group contains two nodes:
+     - a primary node pointing to the nearline storage itself.
+     - a secondary node on a regular disk used for storing files too small for nearline.
+
+    Optional io_config keys:
+     - threshold : integer
+            The smallfile threshold, in bytes.  Files above this size are sent to
+            nearline and put on tape.  The small files are sent to the secondary
+            node.  Default is 1000000000 bytes (1 GB).
+    """
+
+    # SETUP
+
+    def __init__(self, group: StorageGroup, config: dict) -> None:
+        super().__init__(group, config)
+
+        self._threshold = self.config.get("threshold", 1000000000)  # bytes
+
+    # HOOKS
+
+    def set_nodes(self, nodes: list[UpdateableNode]) -> list[UpdateableNode]:
+        """Set nodes used during update.
+
+        Identify primary and secondary nodes.  If both nodes aren't online,
+        we cancel the update.
+
+        Parameters
+        ----------
+        nodes : list of UpdateableNodes
+                The local active nodes in this group.  Will never be
+                empty.
+        idle : boolean
+                True if all the `nodes` were idle when the current
+                update loop started.
+
+        Returns
+        -------
+        nodes : list of UpdateableNodes
+                The passed in `nodes` if both nodes were identified.
+
+        Raises
+        ------
+        ValueError
+            At least one node couldn't be identified, or more than
+            two nodes were provided.
+        """
+        if len(nodes) != 2:
+            raise ValueError(
+                f"need exactly two nodes in StorageGroup group {self.group.name} (have {len(nodes)})"
+            )
+            return False
+
+        if nodes[0].db.io_class == "Nearline":
```

Comment:
I'm assuming here that `smallfile` can be any io class?

Reply:
Yeah, though probably should check that it's not also a Nearline node (since then there's no way to figure out which is which).